xtevent/_eventols.ado

```diff
@@ -159,17 +170,98 @@ program define _eventols, rclass
 		}
 	}
 	else loc indepvars ""
+		
+	******** SA *******
+	if "`sun_abraham'"!=""{
+		* Parse the dependent variable
+		local lhs = "`depenvar'"
+		local rel_time_list = "`included'"
+		
+		* Convert the varlist of relative time indicators to nvarlist 
+		local nvarlist "" // copies of relative time indicators with control cohort set to zero
+		local dvarlist "" // for display
+		foreach l of varlist `rel_time_list' {
+			local dvarlist "`dvarlist' `l'"
+			tempname n`l'
+			qui gen `n`l'' = `l'
+			qui replace `n`l'' = 0 if  `control_cohort' == 1 
+			local nvarlist "`nvarlist' `n`l''" 
+		}	
+
+		* Get cohort count  and count of relative time
+		qui levelsof `cohort' if  `control_cohort' == 0, local(cohort_list) 
+		local nrel_times: word count `nvarlist' 
+		local ncohort: word count `cohort_list'  
+		
+		**** step 2 
+		* Initiate empty matrix for weights 
+		tempname bb ff_w
+		
+		* Loop over cohort and get cohort shares for relative times
+		local nresidlist ""
+		foreach yy of local cohort_list {
+			tempvar cohort_ind resid`yy'
+			qui gen `cohort_ind'  = (`cohort' == `yy') 
+			qui regress `cohort_ind' `nvarlist'  if `touse' & `control_cohort' == 0 [`weight'`exp']  , nocons
+			mat `bb' = e(b) 
+			matrix `ff_w'  = nullmat(`ff_w') \ `bb' 
+			*di "`yy'"
+			*mat li `ff_w'
+			qui predict double `resid`yy'', resid 
+			local nresidlist "`nresidlist' `resid`yy''" //list of variables of residuals 
+		}
+		
+		* Get VCV estimate for the cohort shares using avar
+		* In case users have not set relative time indicators to zero for control cohort
+		* Manually restrict the sample to non-control cohort
+		tempname XX Sxx Sxxi S KSxxi Sigma_ff
+		qui mat accum `XX' = `nvarlist' if  `touse' & `control_cohort' == 0 [`weight'`exp'], nocons
+		mat `Sxx' = `XX'*1/r(N) 
+		mat `Sxxi' = syminv(`Sxx') 
+		qui avar (`nresidlist') (`nvarlist')  if `touse' & `control_cohort' == 0 [`weight'`exp'], nocons robust
+		mat `S' = r(S) 
+		mat `KSxxi' = I(`ncohort')#`Sxxi' 
+		mat `Sigma_ff' = `KSxxi'*`S'*`KSxxi'*1/r(N) 
+		// Note that the normalization is slightly different from the paper
+		// The scaling factor is 1/N for N the obs of cross-sectional units
+		// But here estimates are on the panel, which is why it is 1/NT instead
+		// Should cancel out for balanced panel, but unbalanced panel is a TODO
+		
+		**** step 1 
+		* Prepare interaction terms for the interacted regression
+		local cohort_rel_varlist "" // hold the temp varnames	
+		foreach l of varlist `rel_time_list' { 
+			foreach yy of local cohort_list {  
+				tempvar n`n`l''_`yy'
+				qui gen `n`n`l''_`yy''  = (`cohort' == `yy') * `n`l' '
+				// TODO: might be more efficient to use the c. operator if format w/o missing
+				local cohort_rel_varlist "`cohort_rel_varlist' `n`n`l''_`yy''"
+				if "`saveint'"!=""{
+					loc lnumber=substr("`l'",7,2)
+					qui gen _interact_`lnumber'_c`yy' = `n`n`l''_`yy''

```

Comment:
@Constantino-Carreto-Romero This hardcoded substring length makes me nervous, how are we sure that this will always be the proper length?

(I am aware this is SA's code, but might as well check it)

Reply:
I added that code for the interaction variables to be saved when the suboption `saveinteract` is specified. It substrings on the _k* variables generated in `_eventgenvars` and whose names are stored in the local `included`. The _k* are for instance `_k_eq_m4 _k_eq_m3 _k_eq_m2 _k_eq_p0 _k_eq_p1 _k_eq_p2 _k_eq_p3 _k_eq_p4`, so it will substring `m4 m3 ... p3 p4`. In that code stage, the _k* variables will always have that name structure . It can only be modified later in `_eventols` with an added prefix by the option `savek`.